Assets/MixedRealityToolkit.Services/InputSimulation/Editor/BaseInputSimulationService.cs

```diff
@@ -0,0 +1,155 @@
+            return activeControllers;
+        }
+
+        #region BaseInputDeviceManager Implementation
+
+        public BaseInputSimulationService(
+            IMixedRealityServiceRegistrar registrar,
+            IMixedRealityInputSystem inputSystem,
+            string name,
+            uint priority,
+            BaseMixedRealityProfile profile) : base(registrar, inputSystem, name, priority, profile)
+        {}
+
+        #endregion BaseInputDeviceManager Implementation
+
+        // Register input sources for hands based on hand data
+        protected void UpdateHandDevice(HandSimulationMode simulationMode, Handedness handedness, SimulatedHandData handData)
+        {
+            if (handData != null && handData.IsTracked)
+            {
+                SimulatedHand controller = GetOrAddHandDevice(handedness, simulationMode);
+                controller.UpdateState(handData);
+            }
+            else
+            {
+                RemoveHandDevice(handedness);
+            }
+        }
+
+        public SimulatedHand GetHandDevice(Handedness handedness)
+        {
+            if (trackedHands.TryGetValue(handedness, out SimulatedHand controller))
+            {
+                return controller;
+            }
+            return null;
+        }
+
+        protected SimulatedHand GetOrAddHandDevice(Handedness handedness, HandSimulationMode simulationMode)
+        {
+            var controller = GetHandDevice(handedness);
+            if (controller != null)
+            {
+                if (controller.SimulationMode == simulationMode)
+                {
+                    return controller;
+                }
+                else
+                {
+                    // Remove and recreate hand device if simulation mode doesn't match
+                    RemoveHandDevice(handedness);
+                }
+            }
+
+            SupportedControllerType st = simulationMode == HandSimulationMode.Gestures ? SupportedControllerType.GGVHand : SupportedControllerType.ArticulatedHand;
+            IMixedRealityPointer[] pointers = RequestPointers(st, handedness);
+
+            var inputSource = InputSystem?.RequestNewGenericInputSource($"{handedness} Hand", pointers, InputSourceType.Hand);
+            switch (simulationMode)
+            {
+                case HandSimulationMode.Articulated:
+                    controller = new SimulatedArticulatedHand(TrackingState.Tracked, handedness, inputSource);
+                    break;
+                case HandSimulationMode.Gestures:
+                    controller = new SimulatedGestureHand(TrackingState.Tracked, handedness, inputSource);
+                    break;
+                default:
+                    controller = null;
+                    break;
+            }
+
+            System.Type controllerType = simulationMode == HandSimulationMode.Gestures ? typeof(SimulatedGestureHand) : typeof(SimulatedArticulatedHand);
+            if (controller == null)
```

Comment:
>controller [](start = 16, length = 10)

I think this should be controllerType (judging by how Debug.LogError is mentioning controllerType

Reply:
Ah, I guess maybe the line before was just for creating a more debug-friendly message. I guess in this case it was weird to read a set and then a null check for something that wasn't just set.

---
In reply to: [314510681](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5173#discussion_r314510681) [](ancestors = 314510681)

Reply:
Yeah nevermind :)

---
In reply to: [314511235](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5173#discussion_r314511235) [](ancestors = 314511235,314510681)